### ISSUE TITLE
Explicit keyword for interface contracts

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#847]
Splitting out this discussion from https://github.com/ethereum/solidity/issues/649.

I think it would make sense having interfaces explicitly defined in Solidity source code:

```
interface Token {
  function balance(address owner) returns (uint);
  function transfer(address recipient, uint amount);
}
```

Compared to _abstract contracts_, none of the interface methods can have a body.

**Comment by chriseth** [comment]
Please list pros and cons

**Comment by chriseth** [comment]
What about:

Warn if contract is abstract but not "marked" as abstract.
You can mark a contract as abstract by using  `contract a is abstract {}`.
It is a bit weird, because `abstract` would be a keyword that is not inherited, i.e. contracts that derive from `a` will not necessarily be abstract.

**Comment by axic** [comment]
I think I've explained the pros since in #649.

Both `contract interface Token {}` and `contract Token interface { }` could work, but mixing it with inheritance (the `is` keyword) is dangerous.

**Comment by VoR0220** [comment]
I think interface could be its own type and we can run it like in golang. Essentially what would happen is that you can only apply function types to interface types but structs can inherit an interface. This interface can then easily be passed around with different structs that inherit or have the same functions as the interface. You should then be able to assert from the interface what type of object it is. The null interface could be a var and passed in as a generic.

**Comment by VoR0220** [comment]
Only problem is then whether or not structs implement their own version of the function. Which they probably shouldn't. However it could be mandated to check for all types defined in functions that take in an interface type. Perhaps contracts themselves could be passed in as types from this point as well so long as the functions are implemented in the contract body.

**Comment by axic** [comment]
One more idea to interfaces: they could be made pure interfaces by not allowing inheritance. That ensures an interface remains self-contained and hopefully limited in size.

**Comment by VoR0220** [comment]
I'm not sure how we do that...how do we limit inheritance?

**Comment by esques** [comment]
While doing OOA on a contract, I noticed that I miss the notion of interface in the language.
In object orientation, interfaces are used to designate roles to a class (ERC20 token, Iterable, Stoppable etc.) . An interface **does not ** equal to an abstract class. It just defines the signature of a bundle that must be implemented by the class {public members, public functions, public events}. 

I think interfaces are both simple to implement and maintain and very useful. 
Because sane multiple inheritance is impossible, they help to define multiple aspects of an object.
I would really plea for implementing them.

**Comment by chriseth** [comment]
We have a tradeoff here: We can make the language more complicated by adding more notions and keywords (I think C# is a negative example here). On the other hand, this complication prevents people from making mistakes by accidentally implementing a function in a contract that is meant as a mere interface. Since these complications come in small steps, I'm not sure many people would argue against a single step until it is "too late". What are your opinions here, @rolandkofler @axic @VoR0220 ?

**Comment by VoR0220** [comment]
An interface should be its own type. Plain and simple. The notion of abstract contracts should be disintegrated in future releases and replaced with an interface type. If a contract or a struct for that matter (have to get extra specific for struct methods) implements all of that interfaces methods it can be passed into a function signature containing said interface. The way that it currently is set up is too wide open and invites errors. We simply need stronger types. This is my opinion but what benefit do we get from the current abstract function set in terms of implementing this kind of a feature? It hinders us more than helps us imo.

**Comment by axic** [comment]
I still think pure interfaces (which cannot inherit other interfaces) are very useful to define a limited set of mandatory public signatures that must be present. These in our current state can only be defined as functions in the interface, but the implementor might choose a public accessor to fulfil the requirements.

I am not sure whether we should include events in that or not, given the special nature of events.

I think most cases of abstract contracts can be eliminated by templates.

**Comment by VoR0220** [comment]
WRT events...I'm against it for now...treat it like libraries and inheritance. Keep a restriction now...lift it up with time if the community demands it.

**Comment by esques** [comment]
> if the community demands it.

Interfaces would really help, they are just an enforcement of strong typification as @axic  said, help greatly promoting a modular ecosystem, and have virtually no security risk or performance impact or consequences on the functionality. They just help keep up a good domain model.

**Comment by VoR0220** [comment]
@rolandkofler I'm agreeing with you. I'm saying specifically wrt @axic 's talk about including events in the interface type as functions that must be present within the contract. IMO it makes the type more complicated and will make the development of such a type longer and harder to produce.

**Comment by esques** [comment]
@VoR0220 yes, fully agree. Start with a *minimal viable* interface. Plain, no inheritance, no member variables, just functions.

**Comment by VoR0220** [comment]
@axic Why not just add the `pure` as a type modifier to type `interface`...this means that it can't inherit other interfaces. An example of what I imagine as an interface type would be like such:

library of interfaces
```
library myInterfaces {
    pure interface Animal { //this means the interface cannot be inherited once passed in as this interface
         Speak() returns (string);
    }

   interface Earthling {
         IsEarthling() returns (bool);
   } 
   interface Human { //this interface can be inherited from once passed in as is. 
        Name() returns (string);
   }

}
```

contract Kitty:

```
contract Kitty {
     function Speak() returns (string) { return "meow"; }

     function IsEarthling() returns (string) { return true; } 

     function Name() returns (string) { return "mittens";}
     
}
```

contract Bob:

```
contract BobTheHuman {
    function Speak() returns (string) { return "Hi!"; }

    function Name() returns (string) { return "Bob"; }

    function IsEarthling() returns (bool) { return true; }

}
```

contract Main: 
```
contract Main {
BobTheHuman bob;
Kitty cat;

// this will work because Bob can be turned into a human and also turned into an earthling 
// because his interfaces are normal and flexible
function assertBobIsAnEarthling() returns (bool) {
     return IsHumanEarthling(bob);
}

function IsHumanEarthling(Human _human) returns (bool) {
    return LifeFormIsEarthling(_human);
}

function LifeFormIsEarthling(Earthling _lifeform) returns (bool) {
    return _lifeform.IsEarthling();
}
// the following will not work because kitty cat cannot be turned into anything but an animal, and will
// fail the second that we try to  turn kitty cat into an earthling
function assertKittyCatIsAnEarthling() returns (bool) {
      return IsAnimalEarthling(cat);
}

function isAnimalEarthling(Animal _animal) returns (bool) {
     return LifeFormIsEarthling(_animal); //fail right here.
}


}

```

**Comment by esques** [comment]
@VoR0220  `pure` is the wrong keyword IMHO, `simple` or even better `naive` would do it.

**Comment by VoR0220** [comment]
I would assume `simple` would be the default interface and would enable flexibility...perhaps `strict` would be a better term for the above?

**Comment by axic** [comment]
@VoR0220 in the examples above `Kitty` and `BobTheHuman` didn't inherit from the interface.

My current criteria:
- interface is defined by the `interface` keyword
- cannot inherit
- cannot have variables defined
- cannot have function bodies
- cannot have structs defined (though this might make sense to lift later on - it is easier to start with a limited subset and enable more features as confidence grows, than the doing reverse)

Made the changes and it is only ~50 lines of code to solc. (Available here: https://github.com/ethereum/solidity/tree/interface-keyword)

This compiles:
```
interface Token {
        event Transfer(address indexed _from, address indexed _to, uint256 _value);
        event Approval(address indexed _owner, address indexed _spender, uint256 _value);

        function totalSupply() constant returns (uint256 supply);
        function balanceOf(address _owner) constant returns (uint256 balance);
        function transfer(address _to, uint256 _value) returns (bool success);
        function transferFrom(address _from, address _to, uint256 _value) returns (bool success);
        function approve(address _spender, uint256 _value) returns (bool success);
        function allowance(address _owner, address _spender) constant returns (uint256 remaining);
}
```

This essentially maps to the fact that what can be expressed clearly in the  JSON ABI can be transformed into an interface "contract" and vice versa.

**Comment by VoR0220** [comment]
@axic this is definitely the easier way to supply it I will agree there, but there are some problems I have in terms of checking of contracts being passed in with certain types...I would like to get an interface in a form similar to how they interact in Golang (and I know that that is a stretch but I think it could definitely help out safety in many ways if done properly) see here: 

http://jordanorelli.com/post/32665860244/how-to-use-interfaces-in-go

Particularly as it relates to the empty interface I suppose that wouldn't fit here. However, in terms of reflecting and changing into different types based on whether they implement an interface, I think this is very useful in the mindset of contracts and enables contract safety in a way because we can check via a hash of the code and whether or not this code's hash fits this type being passed in, and can be done via a switch case.

**Comment by axic** [comment]
Without having read the entire article, it starts with:

> This is a core concept in Go’s type system; instead of designing our abstractions in terms of what kind of data our types can hold, we design our abstractions in terms of what actions our types can execute.

Which the above proposal matches 100% 😉 

I don't think automatic interface lookup works well in our scenario. One of the main points of the `interface` type is to have a compile type verification that your output contract satisfies a required ABI, because in our case, updating a contract is tad more problematic (at least today) than installing a new version of a desktop software.

> because we can check via a hash of the code and whether or not this code's hash fits this type being passed in

I don't fully follow here.

**Comment by VoR0220** [comment]
@axic What I mean is that I should be able to pass an interface into a function signature as a type and then check to see if its the particular interface I want. A pseudocode example of how this would work is the following:

```
assume that A = Hash(fileA.sol:Contract A), B = Hash(fileB.sol:Contract B), C = Hash(fileC.sol:Contract C)

function (interface _myInterface) {
switch _myInterface.type {
case A: //check if the incoming contract subscribes to hash of A, if so continue
// do stuff with A
case B: // ibid hash of B
// do stuff with B
case C: // ibid hash of C
// do stuff with C
default:
throw;
}

```

**Comment by pirapira** [comment]
Can this be closed now that #1688 has been merged?

**Comment by axic** [comment]
Yes!
